managed/services/telemetry/telemetry.go

```diff
@@ -160,7 +161,60 @@ func (s *Service) DistributionMethod() serverpb.DistributionMethod {
 	return s.sDistributionMethod
 }
 
-func (s *Service) prepareReport(ctx context.Context) *reporter.ReportRequest {
+func (s *Service) processSendCh(ctx context.Context) {
+	var reportsSync sync.Mutex
+	var inflightReports []*pmmv1.ServerMetric
+	var sendCtx context.Context
+	var cancel context.CancelFunc
+
+	cleanup:= func() {
+		reportsSync.Lock()
+		inflightReports = nil
+		sendCtx = nil
+		cancel = nil
+		reportsSync.Unlock()
+	}
+
+	for {
+		select {
+		case report, ok := <-s.sendCh:
+			if ok {
+				s.l.Debug("Processing telemetry report.")
+				reportsSync.Lock()
+				inflightReports = append(inflightReports, report)
+				if sendCtx != nil {
+					cancel()
+				}
+				sendCtx, cancel = context.WithCancel(ctx)
+
+				reportsCopy := make([]*pmmv1.ServerMetric, len(inflightReports))
+				copy(reportsCopy, inflightReports)
+				reportsSync.Unlock()
+
+				go func(ctx context.Context, reports []*pmmv1.ServerMetric, cleanup func()) {
+					err := s.send(ctx, &reporter.ReportRequest{
+						Metrics: reports,
+					})
+					if err != nil {
+						s.l.Debugf("Telemetry info not sent, due to error: %s.", err)
+						return
+					}
+
+					cleanup()
+
+					s.l.Debug("Telemetry info sent.")
+				}(sendCtx, reportsCopy, cleanup)
```

Comment:
do we really need to pass these arguments, they are all available in a function and don't save us from a race condition.

Reply:
I know it can be a closure, but I think it is better to explicitly specify what is used, and it is personal preference for cases where we have concurrency. Please let me know if you think it is better to use varibles directly.